Deploy zip and npm packages.

mramato opened the pull request:
This change allows travis to deploy the generate zip files and npm packages in addition to the build output.  Each package gets a unique name based on branch name and build number.  Also updated deployment step not to delete these packages so that they can be depended on until the branch is merged.

I had to open a PR just to test this change, so if everything works, you'll notice 3 items in the status below.

mramato commented:
@ggetz can you please review this.

pjcozzi commented:
Can you update the release guide?

mramato commented:
> Can you update the release guide?

More work/research needs to be done to use this for the release process.  The reason is that we would need to generate the GitHub tag as well as actually deploy the npm package.  This requires the person doing the release to build locally anyway.

mramato commented:
So apparently TRAVIS_BRANCH is not the branch names for PRs but is instead the branch being PRed into, so I'll definitely need to make some updates here before merging.

ggetz commented on gulpfile.js:
```diff
@@ -464,12 +466,19 @@ function deployCesium(bucketName, uploadDirectory, cacheControl, done) {
+            var objectToDelete = [];
+            existingBlobs.forEach(function(file) {
+                //Don't delete generate zip files.
+                if (!/\.(zip|tgz)$/) {
```
The delete request failed when running in travis. I think this regex should be calling `test` on  `file`.

mramato replied:
Whoop, thanks.

mramato commented:
@tfili as you requested this morning, you can now reference `http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/deploy-zips/cesium-1.19.0-deploy-zips.15029.tgz` (obviously the filename changes based on branch) and that will allow you to use a dev branch as a node dependency.

We aren't cleaning up these zips yet, but when we do they should still be around for several months at a time so them going away should never be an issue.  Obviously these shouldn't be used for production deployment.

mramato commented:
Okay, I fixed the issues and squashed the commits; so this is good to go when all is green.

I'll also add (since I didn't know) that if you go to the [branches page](https://github.com/AnalyticalGraphicsInc/cesium/branches) and click on the green checkmark, you can access the same deployment urls for branches that don't have PRs yet.

mramato commented:
This is ready.

ggetz commented:
Can we make the message "Deploying x..." change to something like "x has been deployed." once the process is complete? That's what the Travis statuses do.

mramato commented:
Sure, done.  I just changed it to `Waiting for build` and then `Deployed`, I was going to put urls in there, but GitHub didn't like it.

ggetz commented on .travis.yml:
```diff
@@ -17,12 +22,13 @@ script:
   - npm run deploy-s3 -- -b cesium-dev -d cesium/$TRAVIS_BRANCH --confirm -c 'no-cache'
-  - npm run deploy-status-success
+  - npm run deploy-status -- --status success
+  - npm run deploy-status -- --status success --message Deployed
```
Is this command supposed to be run twice?

mramato replied:
No, it's Monday and I haven't had enough caffeine yet.  Thanks

mramato commented:
Okay, hopefully once this turns green it's really good to go.  Thanks again @ggetz

mramato commented:
This is ready, the build is failing because of #3757, but the deployment is working fine.

ggetz commented:
Looks good, thanks @mramato !

mramato commented:
@pjcozzi can you please merge this.
